controllers/lvm_volumegroup.go

```diff
@@ -0,0 +1,126 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	lvmv1alpha1 "github.com/red-hat-storage/lvm-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	lvmvgName = "lvmvg-manager"
+)
+
+type lvmVG struct{}
+
+func (c lvmVG) getName() string {
+	return lvmvgName
+}
+
+func (c lvmVG) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+
+	lvmVolumeGroups, err := c.getLvmVolumeGroups(r, lvmCluster)
+	if err != nil {
+		return err
+	}
+
+	for _, volumeGroup := range lvmVolumeGroups {
+		result, err := cutil.CreateOrUpdate(ctx, r.Client, volumeGroup, func() error {
+			// no need to mutate any field
+			return nil
+		})
+
+		if err != nil {
+			r.Log.Error(err, "LvmVolumeGroup reconcile failure", "name", volumeGroup.Name)
+		} else {
+			r.Log.Info("LvmVolumeGroup", "operation", result, "name", volumeGroup.Name)
+		}
+
+	}
+	return nil
+}
+
+func (c lvmVG) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+
+	existingLvmVg := &lvmv1alpha1.LVMVolumeGroup{}
+	vgcrs, err := c.getLvmVolumeGroups(r, lvmCluster)
+	if err != nil {
+		return err
+	}
+
+	for _, vgcr := range vgcrs {
+		err := r.Client.Get(ctx, types.NamespacedName{Name: vgcr.Name, Namespace: vgcr.Namespace}, existingLvmVg)
+
+		if err != nil {
+			// already deleted in previous reconcile
+			if errors.IsNotFound(err) {
+				r.Log.Info("LVMVG deleted", "Name", vgcr.Name)
+				return nil
+			}
+			r.Log.Error(err, "failed to retrieve LVMVG resource", "Name", vgcr.Name)
+			return err
+		}
+
+		// if not deleted, initiate deletion
+		if existingLvmVg.GetDeletionTimestamp().IsZero() {
+			if err = r.Client.Delete(ctx, existingLvmVg); err != nil {
+				r.Log.Error(err, "failed to delete LVMVG", "Name", existingLvmVg.Name)
+				return err
+			} else {
+				r.Log.Info("initiated LVMVG deletion", "Name", existingLvmVg.Name)
+			}
+		} else {
+			// set deletion in-progress for next reconcile to confirm deletion
+			return fmt.Errorf("LVMVG %s is already marked for deletion", existingLvmVg.Name)
+		}
+
+	}
+	return nil
+}
+
+func (c lvmVG) updateStatus(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	// intentionally empty
+	return nil
+}
+
+func (c lvmVG) getLvmVolumeGroups(r *LVMClusterReconciler, instance *lvmv1alpha1.LVMCluster) ([]*lvmv1alpha1.LVMVolumeGroup, error) {
```

Comment:
`getLvmVolumeGroups` does not return any error. Should remove `error` from the return types.

Reply:
done.